File: 2023/javascript/day04a/day04a.js
const fs = require('fs');

function usage() {
    console.log(`usage: ${process.argv[0]} ${process.argv[1]} <file>`);
    process.exit(1);
}

function processInput(contents) {
    let lines = contents.trim().split('\n');
    let result = 0;
    for ( let line of lines ) {
        let rest = line.split(': ')[1];
        let winningStr = rest.split(' | ')[0];
        let winningSet = new Set();
        for ( let numStr of winningStr.trim().split(' ') ) {
            if ( numStr != '' ) {
                let num = parseInt(numStr);
                winningSet.add(num);
            }
        }
        let handStr = rest.split(' | ')[1];
        let handSet = new Set();
        for ( let numStr of handStr.trim().split(' ') ) {
            if ( numStr != '' ) {
                let num = parseInt(numStr);
                handSet.add(num);
            }
        }
        let intersection = new Set();
        for ( let num of winningSet ) {
            if ( handSet.has(num) ) {
                intersection.add(num);
            }
        }
        let count = intersection.size;
        if ( count > 0 ) {
            result += 1 << (count - 1);
        }
    }
    return result;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    let filename = process.argv[2];
    let contents = fs.readFileSync(filename, 'utf8');
    let result = processInput(contents);
    console.log(`result = ${result}`);
}

main()

File: 2023/kotlin/day04a/day04a.kt
import java.io.File

fun usage() {
    println("usage: java -jar <jarfile> <infile>")
    System.exit(1)
}

fun process(contents: String): Int {
    var result = 0
    for ( line in contents.lines() ) {
        if ( line.isEmpty() ) {
            break
        }
        line.split(Regex(""":\s+""")).let { (_, rest) ->
            rest.split(Regex("""\s+\|\s+""")).let { (winningStr, handStr) ->
                var winningSet = HashSet<Int>()
                for ( numStr in winningStr.split(Regex("""\s+""")) ) {
                    val num = numStr.toIntOrNull()!!
                    winningSet.add(num);
                }
                var handSet = HashSet<Int>()
                for ( numStr in handStr.split(Regex("""\s+""")) ) {
                    val num = numStr.toIntOrNull()!!
                    handSet.add(num);
                }
                val intersection = winningSet.intersect(handSet)
                val count = intersection.size
                if ( count > 0 ) {
                    result += 1 shl (count - 1);
                }
            }
        }
    }
    return result
}

fun main(args: Array<String>) {
    if ( args.size < 1 ) {
        usage()
    }
    val filename = args[0]
    val contents = File(filename).readText()
    val result = process(contents)
    println("result = $result")
}
